sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -211,6 +211,7 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    */
   def load(path: String): DataFrame = {
     // force invocation of `load(...varargs...)`
+    removeExtraOptionsByKey("path")
```

Comment:
I don't mind doing separately if this is going to be more controversial but I think we should just use `collection.mutable.LinkedHashMap` and `collection.immutable.ListMap`.

I think it's very unlikely that users depend on the indeterministic behaviours of `option("paTh").option("Path")` and it sounds more like a bug fix to me - the same option `path` is set later via `option` but not respected. Took a cursory look and seems feasible without a big change.

Reply:
+1. I don't think an "indeterministic behavior" should be considered in backward compatibility. We should just use `LinkedHashMap`, to solve this problem completely.

Reply:
Got it. Thank you, @HyukjinKwon and @cloud-fan .

Reply:
FYI, `collection.mutable.LinkedHashMap` turns out unsafe because `.toMap` operation lose the orderness and converts it to `HashTrieMap`.

Reply:
For a POC, I'll create a separate PR, @cloud-fan and @HyukjinKwon .

Reply:
Before going to the new PR, the following is the result on `LinkedHashMap`. `LinkedHashMap` doesn't solve the problem.
```scala
val extraOptions = new scala.collection.mutable.LinkedHashMap[String, String]
extraOptions += ("paTh" -> "1")
extraOptions += ("PATH" -> "2")
extraOptions += ("Path" -> "3")
extraOptions += ("patH" -> "4")
extraOptions += ("path" -> "5")
extraOptions.toMap

// Exiting paste mode, now interpreting.

extraOptions: scala.collection.mutable.LinkedHashMap[String,String] = Map(paTh -> 1, PATH -> 2, Path -> 3, patH -> 4, path -> 5)
res0: scala.collection.immutable.Map[String,String] = Map(PATH -> 2, path -> 5, patH -> 4, Path -> 3, paTh -> 1)
```

Reply:
I'm testing the PR.

Reply:
I redesigned the PR.